readme.txt
=== Price Based on Country for WooCommerce ===
Contributors: oscargare
Tags: price based country, dynamic price based country, price by country, dynamic price, woocommerce, geoip, country-targeted pricing
Requires at least: 3.8
Tested up to: 6.1
Stable tag: 3.1.0
License: GPLv2 or later
License URI: http://www.gnu.org/licenses/gpl-2.0.html

Add multicurrency support to WooCommerce, allowing you set product's prices in multiple currencies based on country of your site's visitor.

== Description ==

**Price Based on Country for WooCommerce** allows you to sell the same product in multiple currencies based on the country of the customer.

= How it works =

The plugin detects automatically the country of the website visitor throught the geolocation feature included in WooCommerce (2.3.0 or later) and display the currency and price you have defined previously for this country.

You have two ways to set product's price for each country:

* Calculate price by applying the exchange rate.
* Set price manually.

When country changes on checkout page, the cart, the order preview and all shop are updated to display the correct currency and pricing.

= Multicurrency =
Sell and receive payments in different currencies, reducing the costs of currency conversions.

= Country Switcher =
The extension include a country switcher widget to allow your customer change the country from the frontend of your website.

= Shipping currency conversion =
Apply currency conversion to Flat and International Flat Rate Shipping.

= Compatible with WPML =
WooCommerce Product Price Based on Countries is officially compatible with [WPML](https://wpml.org/extensions/woocommerce-product-price-based-countries/).

= Upgrade to Pro =

>This plugin offers a Pro addon which adds the following features:

>* Guaranteed support by private ticket system.
>* Automatic updates of exchange rates.
>* Add an exchange rate fee.
>* Round to nearest.
>* Display the currency code next to price.
>* Compatible with the WooCommerce built-in CSV importer and exporter.
>* Thousand separator, decimal separator and number of decimals by pricing zone.
>* Currency switcher widget.
>* Support to WooCommerce Subscriptions by Prospress .
>* Support to WooCommerce Product Bundles by SomewhereWarm .
>* Support to WooCommerce Product Add-ons by WooCommerce .
>* Support to WooCommerce Bookings by WooCommerce .
>* Support to WooCommerce Composite Product by SomewhereWarm.
>* Support to WooCommerce Name Your Price by Kathy Darling.
>* Bulk editing of variations princing.
>* Support for manual orders.
>* More features and integrations is coming.

>[Get Price Based on Country Pro now](https://www.pricebasedcountry.com?utm_source=wordpress.org&utm_medium=readme&utm_campaign=Extend)

= Requirements =

* WooCommerce 3.4 or later.
* If you want to receive payments in more of one currency, a payment gateway that supports them.

== Installation ==

1. Download, install and activate the plugin.
1. Go to WooCommerce -> Settings -> Product Price Based on Country and configure as required.
1. Go to the product page and sets the price for the countries you have configured avobe.

= Adding a country selector to the front-end =

Once you’ve added support for multiple country and their currencies, you could display a country selector in the theme. You can display the country selector with a shortcode or as a hook.

**Shortcode**

[wcpbc_country_selector other_countries_text="Other countries"]

**PHP Code**

do_action('wcpbc_manual_country_selector', 'Other countries');

= Customize country selector (only for developers) =

1. Add action "wcpbc_manual_country_selector" to your theme.
1. To customize the country selector:
	1. Create a directory named "woocommerce-product-price-based-on-countries" in your theme directory.
	1. Copy to the directory created avobe the file "country-selector.php" included in the plugin.
	1. Work with this file.

== Frequently Asked Questions ==

= How might I test if the prices are displayed correctly for a given country? =

If you are in a test environment, you can configure the test mode in the setting page.

In a production environment you can use a privacy VPN tools like [TunnelBear](https://www.tunnelbear.com/) or [ZenMate](https://zenmate.com/)

You should do the test in a private browsing window to prevent data stored in the session. Open a private window on [Firefox](https://support.mozilla.org/en-US/kb/private-browsing-use-firefox-without-history#w_how-do-i-open-a-new-private-window) or on [Chrome](https://support.google.com/chromebook/answer/95464?hl=en)

== Screenshots ==

1. Simple to get started with the Geolocation setup wizard.
2. Unlimited price zones.
3. Pricing zone properties.
4. Pricing zone properties (2).
5. Plugin settings.
6. Set the price manually or calculate by the exchange rate.
7. Includes a country selector widget.

== Changelog ==

= 3.1.0 (2023-03-07) =
* Added: Tested up WooCommerce 7.5.
* Fixed: Style issues on admin pages for the Spanish language.
* Tweak: Improvements in the function that detects cache plugins.

= 3.0.2 (2023-01-26) =
* Added: Tested up WooCommerce 7.4.
* Added: "0.10" to the "Reduce the converted price" options.

= 3.0.1 (2023-01-26) =
* Fixed: the "Do not adjust taxes based on location" field always shows unchecked on the pricing zone settings.
* Fixed: Typo on the invalid ID error message.

= 3.0.0 (2023-01-23) =
* Added: Tested up WooCommerce 7.3.
* Added: New setup wizard to help store owners configure the WooCommerce geolocation feature.
* Added: The country and currency switcher widgets now support Flags!
* Added: Admin interface improvements.
* Added: Export and import pricing zones.
* Added: Re-order pricing zones.
* Added: Enable/Disable pricing zones.
* Added: Support for Google Listing and Ads plugin.
* Added: Compatible with the new WooCommerce Cart and Checkout blocks.
* Tweak: New price loading animation

[Take a look at the new features introduced in version 3.0](https://www.pricebasedcountry.com/2023/01/23/price-based-on-country-for-woocommerce-3-0-released/)

[See changelog for all versions](https://plugins.svn.wordpress.org/woocommerce-product-price-based-on-countries/trunk/changelog.txt).

== Upgrade Notice ==

= 3.0 =
<strong>3.0 is a major update</strong>, make a backup before updating.
<span class="wcpbc-li">This update requires WooCommerce 4.0 or higher.</span>
<span class="wcpbc-li">If you are using the Pro version, you must update it to the latest version.</span>

changelog.txt
= 3.1.0 (2023-03-07) =
* Added: Tested up WooCommerce 7.5.
* Fixed: Style issues on admin pages for the Spanish language.
* Tweak: Improvements in the function that detects cache plugins.

= 3.0.2 (2023-01-26) =
* Added: Tested up WooCommerce 7.4.
* Added: "0.10" to the "Reduce the converted price" options.

= 3.0.1 (2023-01-26) =
* Fixed: the "Do not adjust taxes based on location" field always shows unchecked on the pricing zone settings.
* Fixed: Typo on the invalid ID error message.

= 3.0.0 (2023-01-23) =
* Added: Tested up WooCommerce 7.3.
* Added: New setup wizard to help store owners configure the WooCommerce geolocation feature.
* Added: The country and currency switcher widgets now support Flags!
* Added: Admin interface improvements.
* Added: Export and import pricing zones.
* Added: Re-order pricing zones.
* Added: Enable/Disable pricing zones.
* Added: Support for Google Listing and Ads plugin.
* Added: Compatible with the new WooCommerce Cart and Checkout blocks.
* Tweak: New price loading animation.

= 2.2.9 (2022-11-24) =
* Tweak: PayPal Payments by WooCommerce" 2.0 compatiblity improvements.

= 2.2.8 (2022-11-21) =
* Fixed: Compatibility issue with "PayPal Payments by WooCommerce" 2.0+.

= 2.2.7 (2022-11-09) =
* Added: Tested up WooCommerce 7.1.
* Added: Declare incompatibility with WooCommerce HPOS feature.
* Fixed: "Unexpected currency from order API call" error on "PayPal Payments" after country changing on the checkout.

= 2.2.6 (2022-10-25) =
* Added: Tested up WooCommerce 7.0.
* Fixed: "Illegal mix of collations" error on Analytics queries when the table charset differs from the default server charset.

= 2.2.5 (2022-09-01) =
* Added: Tested up WooCommerce 6.9.
* Fixed: Using remove_filter( "the_posts",... ) in a "the_posts" callback causes the following "the_posts" callback not to be executed.

= 2.2.4 (2022-08-05) =
* Added: Tested up WooCommerce 6.8.
* Fixed: Minor bugs.

= 2.2.3 (2022-05-25) =
* Added: Tested up WooCommerce 6.5.
* Added: Tested up WordPress 6.0.

= 2.2.2 (2022-03-23) =
* Fixed: The pricing fields are visible on the variable products when the "WooCommerce Subscriptions" plugin is enabled.
* Fixed: The net sales amount of the Products Report in Analytics is not converted to the base currency.
* Tweak: Copy the order metadata to the refund orders to improve the accuracy of the reports.

= 2.2.1 (2022-03-03) =
* Added: Tested up WooCommerce 6.3.
* Added: New filter wc_price_based_country_ajax_geo_skip_wrapper.
* Fixed: "Load product price in the background" option does not work if "Discount Rules by Flycart" is enabled.
* Fixed: Pricing zone settings does not display is "Caldera Forms" plugin is enabled.

= 2.2.0 (2022-02-07) =
* Added: Tested up WooCommerce 6.2.
* Fixed: Compatibility issue with PayPal Payments plugin.

= 2.1.1 (2022-01-12) =
* Added: Tested up WooCommerce 6.1.
* Fixed: PHP Warning "function is_ajax is deprecated since WooCommerce 6.1".

= 2.1.0 (2021-12-15) =
* Added: Tested up WooCommerce 6.0.
* Tweak: Update order meta on woocommerce_checkout_update_order_meta hook insteand of woocommerce_new_order to prevent issues.
* Deprecated: WCPBC_Pricing_Zone::set_zone_id
* Deprecated: WCPBC_Pricing_Zone::get_zone_id

= 2.0.30 (2021-11-29) =
* Fixed: PHP error on creating a new order when the second parameter of the 'woocommerce_new_order' hook is missing.
* Fixed: "Load product price in the background" does not work with the Porto theme "Show Skeleton Screens" option on variable products.

= 2.0.29 (2021-11-23) =
* Fixed: PHP warning on shipping cost currency conversion when the amount is zero.
* Fixed: Incorrect decimal separator after saving variations when WPML is active.
* Tweak: "Load product price in the background" compatible with the Porto theme "Show Skeleton Screens" option.

= 2.0.28 (2021-11-09) =
* Added: Tested up WooCommerce 5.9.
* Fixed: Duplicate _wcpbc_base_exchange_rate postmeta on subscription renewal orders.
* Fixed: "Google product feed by Ademti Software" - Incorrect taxes when the country does not match in a pricing zone.
* Tweak: "Load product price in the background" compatible with themes that use the pjax jquery plugin.

= 2.0.27 (2021-10-21) =
* Added: Tested up WooCommerce 5.8.
* Fixed: "Load product price in the background" overwrites texts added by other plugins to the price of variations.
* Fixed: Calculating totals on AJAX requests causes compatibility issues with other plugins (Smart coupons).

= 2.0.26 (2021-10-05) =
* Fixed: PHP error on some cron executions.

= 2.0.25 (2021-09-15) =
* Tweak: Update compatibility with UPS and Fedex shipping plugins.
* Fixed: Incorrect taxes on the Google product feed generated by the "Google product feed by Ademti Software" plugin.

= 2.0.24 (2021-09-07) =
* Added: Tested up WooCommerce 5.7.
* Fixed: Filter by price does not work when other plugin run a query before WordPress runs the main query.
* Fixed: Cart subtotal is not refreshed after user login in WordPress dashboard.
* Fixed: Add aria-label to the country switcher template to meet web accessibility rules.

= 2.0.23 (2021-08-10) =
* Added: Tested up WooCommerce 5.6.
* Added: Tested up WordPress 5.8.
* Tweak: Update compatibility with All Products for WooCommerce Subscriptions.

= 2.0.22 (2021-07-12) =
* Added: Tested up WooCommerce 5.5.
* Fixed: Layout issues on the price fields for variations on large resolutions.
* Fixed: Do not load the "zone pricing" on the Rest API calls of the WooCommerce app.

= 2.0.21 (2021-06-17) =
* Added: Tested up WooCommerce 5.4.
* Fixed: Do not display exchange rates lower than 0.001 using scientific notation.
* Fixed: "Load product price in the background" compatible with FacetWP filters.
* Fixed: Check the function wcpbc_the_zone returns a pricing zone object.

= 2.0.20 (2021-05-18) =
* Fixed: Bug on the filter by price widget.
* Fixed: Duplicate field on the country switcher widget.

= 2.0.19 (2021-05-13) =
* Added: Tested up WooCommerce 5.3.

woocommerce-product-price-based-on-countries.php
<?php
/**
 * Plugin Name: WooCommerce Price Based on Country
 * Plugin URI: https://wordpress.org/plugins/woocommerce-product-price-based-on-countries/
 * Description: Product Pricing and Currency based on Shopper’s Country for WooCommerce.
 * Author: Oscar Gare
 * Version: 3.1.0
 * Author URI: https://www.linkedin.com/in/oscargare
 * Text Domain: woocommerce-product-price-based-on-countries
 * Domain Path: /languages
 *
 * WC requires at least: 4.0
 * WC tested up to: 7.5
 * License: GPLv2
 *
 * @package WCPBC
 */

/*
This program is free software; you can redistribute it and/or
modify it under the terms of the GNU General Public License
as published by the Free Software Foundation; either version 2
of the License, or (at your option) any later version.

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

You should have received a copy of the GNU General Public License
along with this program; if not, write to the Free Software
Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
*/

if ( ! defined( 'ABSPATH' ) ) {
	exit; // Exit if accessed directly.
}

// Define WCPBC_PRO_PLUGIN_FILE.
if ( ! defined( 'WCPBC_PLUGIN_FILE' ) ) {
	define( 'WCPBC_PLUGIN_FILE', __FILE__ );
}

// Include the main Price Based on Country class.
if ( ! class_exists( 'WC_Product_Price_Based_Country' ) ) {
	include_once dirname( __FILE__ ) . '/includes/class-wc-product-price-based-country.php';
}

/**
 * Declare plugin incompatible with the WooCommerce HPOS feature.
 */
add_action(
	'before_woocommerce_init',
	function() {
		if ( class_exists( '\Automattic\WooCommerce\Utilities\FeaturesUtil' ) ) {
			\Automattic\WooCommerce\Utilities\FeaturesUtil::declare_compatibility( 'custom_order_tables', __FILE__, false );
		}
	}
);

/**
 * Returns the main instance of WC_Product_Price_Based_Country to prevent the need to use globals.
 *
 * @since  1.3.0
 * @return WC_Product_Price_Based_Country
 */
function wcpbc() {
	return WC_Product_Price_Based_Country::instance();
}
wcpbc();


includes/class-wc-product-price-based-country.php
<?php
/**
 * WooCommerce Price Based on Country main class
 *
 * @package WCPBC
 */

defined( 'ABSPATH' ) || exit;

/**
 * WC_Product_Price_Based_Country Class
 */
class WC_Product_Price_Based_Country {

	/**
	 * Product Price Based Country version
	 *
	 * @var string
	 */
	public $version = '3.1.0';

	/**
	 * The front-end pricing zone
	 *
	 * @var WCPBC_Pricing_Zone
	 */
	public $current_zone = null;

	/**
	 * Min WC required version.
	 *
	 * @var string
	 */
	protected $min_wc_version = '4.0';

	/**
	 * Min Pro required version.
	 *
	 * @var string
	 */
	protected $min_pro_version = '3.0.0';

	/**
	 * Enviroment alert
	 *
	 * @var string
	 */
	protected $environment_alert = '';

	/**
	 * The single instance of the class.
	 *
	 * @var WC_Product_Price_Based_Country
	 */
	private static $instance = false;

	/**
	 * Main WC_Product_Price_Based_Country Instance
	 *
	 * @static
	 * @see WCPBC()
	 * @return WC_Product_Price_Based_Country
	 */
	public static function instance() {
		if ( empty( self::$instance ) ) {
			self::$instance = new self();
		}
		return self::$instance;
	}

	/**
	 * Get the plugin url.
	 *
	 * @return string
	 */
	public function plugin_url() {
		return plugin_dir_url( WCPBC_PLUGIN_FILE );
	}

	/**
	 * Get the plugin path.
	 *
	 * @return string
	 */
	public function plugin_path() {
		return plugin_dir_path( WCPBC_PLUGIN_FILE );
	}

	/**
	 * Return the plugin base name
	 *
	 * @return string
	 * @since 1.7.4
	 */
	public function plugin_basename() {
		return plugin_basename( WCPBC_PLUGIN_FILE );
	}

	/**
	 * Return the plugin base name
	 *
	 * @return string
	 * @since 3.0.2
	 */
	public function is_pro_compatible() {
		if ( wcpbc_is_pro() ) {
			return isset( WC_Product_Price_Based_Country_Pro::$version ) && version_compare( WC_Product_Price_Based_Country_Pro::$version, $this->min_pro_version, '>=' );
		}
		return true;
	}

	/**
	 * WC_Product_Price_Based_Country Constructor.
	 */
	public function __construct() {
		$this->includes();

		register_activation_hook( WCPBC_PLUGIN_FILE, array( 'WCPBC_Install', 'plugin_activate' ) );
		register_deactivation_hook( WCPBC_PLUGIN_FILE, array( 'WCPBC_Install', 'plugin_deactivate' ) );

		add_action( 'init', array( $this, 'load_textdomain' ) );
		add_filter( 'plugin_action_links_' . plugin_basename( WCPBC_PLUGIN_FILE ), array( $this, 'plugin_action_links' ) );
		add_filter( 'plugin_row_meta', array( $this, 'plugin_row_meta' ), 10, 2 );
		add_action( 'plugins_loaded', array( $this, 'init_plugin' ), 5 );
	}

	/**
	 * Include required files used in admin and on the frontend.
	 */
	private function includes() {

		include_once $this->plugin_path() . 'includes/class-wcpbc-helper-options.php';
		include_once $this->plugin_path() . 'includes/class-wcpbc-pricing-zone.php';
		include_once $this->plugin_path() . 'includes/class-wcpbc-pricing-zones.php';
		include_once $this->plugin_path() . 'includes/wcpbc-core-functions.php';
		include_once $this->plugin_path() . 'includes/class-wcpbc-integrations.php';
		include_once $this->plugin_path() . 'includes/class-wcpbc-frontend.php';
		include_once $this->plugin_path() . 'includes/class-wcpbc-frontend-pricing.php';
		include_once $this->plugin_path() . 'includes/class-wcpbc-ajax-geolocation.php';
		include_once $this->plugin_path() . 'includes/class-wcpbc-update-geoip-db.php';
		include_once $this->plugin_path() . 'includes/admin/class-wcpbc-admin-analytics.php';
		include_once $this->plugin_path() . 'includes/class-wcpbc-product-sync.php';
		include_once $this->plugin_path() . 'includes/class-wcpbc-debug-logger.php';
		include_once $this->plugin_path() . 'includes/class-wcpbc-store-api.php';

		if ( is_admin() || ( defined( 'WP_CLI' ) && WP_CLI ) ) {
			include_once $this->plugin_path() . 'includes/class-wcpbc-install.php';
			include_once $this->plugin_path() . 'includes/admin/class-wcpbc-admin-notices.php';
			include_once $this->plugin_path() . 'includes/admin/class-wcpbc-admin.php';
			include_once $this->plugin_path() . 'includes/admin/class-wcpbc-admin-meta-boxes.php';
			include_once $this->plugin_path() . 'includes/admin/class-wcpbc-admin-ads.php';
			include_once $this->plugin_path() . 'includes/admin/class-wcpbc-setup-wizard.php';
			include_once $this->plugin_path() . 'includes/admin/class-wcpbc-cache-plugins-helper.php';
		}
	}

	/**
	 * Localisation
	 *
	 * @since 1.6.3
	 */
	public function load_textdomain() {
		load_plugin_textdomain( 'woocommerce-product-price-based-on-countries', false, dirname( plugin_basename( WCPBC_PLUGIN_FILE ) ) . '/languages/' );
	}

	/**
	 * Show action links on the plugin screen.
	 *
	 * @since 1.6.11
	 * @param mixed $links Plugin Action links.
	 * @return array
	 */
	public function plugin_action_links( $links ) {

		$action_links = array(
			'settings' => '<a href="' . admin_url( 'admin.php?page=wc-settings&tab=price-based-country' ) . '" aria-label="' . esc_attr__( 'View Price Based on Country settings', 'woocommerce-product-price-based-on-countries' ) . '">' . esc_html__( 'Settings', 'woocommerce-product-price-based-on-countries' ) . '</a>',
		);

		if ( ! wcpbc_is_pro() ) {
			$action_links['get-pro'] = sprintf( '<a target="_blank" rel="noopener noreferrer" style="color:#46b450;" href="%s">%s</a>', wcpbc_home_url( 'plugins-cta-link' ), esc_html__( 'Get Pro', 'woocommerce-product-price-based-on-countries' ) );
		}

		return array_merge( $action_links, $links );
	}

	/**
	 * Show row meta on the plugin screen.
	 *
	 * @since 1.7.0
	 * @param mixed $links Plugin Row Meta.
	 * @param mixed $file  Plugin Base file.
	 * @return array
	 */
	public function plugin_row_meta( $links, $file ) {
		if ( plugin_basename( WCPBC_PLUGIN_FILE ) === $file ) {
			$row_meta = array(
				'docs' => sprintf( '<a target="_blank" rel="noopener noreferrer" href="%s">%s</a>', esc_url( wcpbc_home_url( 'plugins-doc-link', 'docs' ) ), esc_html__( 'Docs', 'woocommerce-product-price-based-on-countries' ) ),
			);

			return array_merge( $links, $row_meta );
		}

		return (array) $links;
	}

	/**
	 * After plugin row. Displays an alert if the Pro version is incompatible.
	 *
	 * @since 3.0.2
	 */
	public function after_plugin_row() {

		$wp_list_table = _get_list_table(
			'WP_Plugins_List_Table',
			array(
				'screen' => get_current_screen(),
			)
		);

		printf(
			'<tr class="plugin-update-tr active" id="price-based-country-pro-incompatible-pro">' .
			'<td colspan="%s" class="plugin-update colspanchange">' .
			'<div class="update-message notice inline notice-error notice-alt"><p>%s <a href="%s">%s</a>.</p></div>' .
			'</td>' .
			'</tr>',
			esc_attr( $wp_list_table->get_column_count() ),
			wp_kses_post( $this->environment_alert ),
			esc_url( admin_url( 'admin.php?page=wc-settings&tab=price-based-country&section=license' ) ),
			esc_html__( 'Check license status', 'woocommerce-product-price-based-on-countries' )
		);

		if ( function_exists( 'wc_enqueue_js' ) ) {
			wc_enqueue_js(
				"$('tr[data-plugin=\"woocommerce-product-price-based-on-countries/woocommerce-product-price-based-on-countries.php\"]').addClass('update');"
			);
		}
	}

	/**
	 * Checks the environment for compatibility problems.
	 *
	 * @return boolean
	 */
	private function check_environment() {
		if ( ! defined( 'WC_VERSION' ) ) {
			// translators: HTML Tags.
			$this->environment_alert = sprintf( __( '%1$sPrice Based on Country%2$s requires WooCommerce to be activated to work. Learn how to install Price Based on Country in the %3$sGetting Started Guide%4$s.', 'woocommerce-product-price-based-on-countries' ), '<strong>', '</strong>', '<a href="https://www.pricebasedcountry.com/docs/getting-started/">', '</a>' );
			return false;
		}

		if ( version_compare( WC_VERSION, $this->min_wc_version, '<' ) ) {
			// translators: HTML Tags.
			$this->environment_alert = sprintf( __( 'The minimum WooCommerce version required for this plugin is %1$s. You are running %2$s.', 'woocommerce-product-price-based-on-countries' ), $this->min_wc_version, WC_VERSION );
			return false;
		}

		if ( ! $this->is_pro_compatible() ) {
			// translators: HTML Tags.
			$this->environment_alert = sprintf( __( 'You are using a deprecated version of the %1$sPrice Based on Country Pro%2$s plugin. All plugin features have been disabled to avoid issues. Please, update Price Based on Country %1$sPro%2$s to the latest version.', 'woocommerce-product-price-based-on-countries' ), '<strong>', '</strong>' );

			if ( class_exists( 'WCPBC_Admin' ) ) {
				// Admin init to allow users set the license.
				WCPBC_Admin::init();

				// Displays warning message after plugin row.
				add_action( 'after_plugin_row_woocommerce-product-price-based-on-countries/woocommerce-product-price-based-on-countries.php', array( $this, 'after_plugin_row' ), 20 );
			}
			return false;
		}

		return true;
	}

	/**
	 * Init plugin
	 *
	 * @since 1.8.0
	 */
	public function init_plugin() {
		if ( ! $this->check_environment() ) {
			add_action( 'admin_notices', array( $this, 'environment_notice' ) );
			return;
		}

		if ( class_exists( 'WCPBC_Install' ) ) {
			// Admin request.
			WCPBC_Install::init();
			WCPBC_Admin::init();
			WCPBC_Admin_Meta_Boxes::init();
			WCPBC_Admin_Notices::init();
			WCPBC_Admin_Ads::init();
			WCPBC_Setup_Wizard::init();
			WCPBC_Cache_Plugins_Helper::init();
		}

		WCPBC_Debug_Logger::init();
		WCPBC_Frontend::init();
		WCPBC_Ajax_Geolocation::init();
		WCPBC_Product_Sync::init();
		WCPBC_Admin_Analytics::init();

		if ( version_compare( WC_VERSION, '3.4', '>=' ) && version_compare( WC_VERSION, '3.9', '<' ) ) {
			WCPBC_Update_GeoIP_DB::init();
		}

		add_action( 'widgets_init', array( $this, 'register_widgets' ) );
		add_action( 'woocommerce_init', array( $this, 'frontend_init' ), 999 );
		add_action( 'init', array( $this, 'ajax_frontend_init' ), 9999 );
	}

	/**
	 * Display the environment alert
	 */
	public function environment_notice() {
		echo '<div id="message" class="error">' . sprintf( '<p><strong>%1$s</strong></p>%2$s', 'Price Based on Country - ' . esc_html__( 'Heads up!', 'woocommerce-product-price-based-on-countries' ), wp_kses_post( wpautop( $this->environment_alert ) ) ) . '</div>';
	}

	/**
	 * Register Widgets
	 *
	 * @since 1.5.0
	 */
	public function register_widgets() {
		if ( class_exists( 'WC_Widget' ) ) {
			include_once $this->plugin_path() . 'includes/class-wcpbc-widget-country-selector.php';
			register_widget( 'WCPBC_Widget_Country_Selector' );
		}
	}

	/**
	 * Init front-end
	 */
	public function frontend_init() {

		if ( $this->is_rest_api_frontend() ) {
			WCPBC_Store_API::init();
		}

		if ( ! $this->is_frontend() || apply_filters( 'wc_price_based_country_stop_pricing', false ) ) {
			// Do only on frontend requests.
			return;
		}

		$this->frontend_pricing_init();
	}

	/**
	 * Frontend pricing init.
	 *
	 * @since 3.0.0
	 */
	public function frontend_pricing_init() {
		if ( did_action( 'wc_price_based_country_before_frontend_init' ) || doing_action( 'wc_price_based_country_before_frontend_init' ) ) {
			// do only once.
			return;
		}

		if ( ! did_action( 'before_woocommerce_init' ) || doing_action( 'before_woocommerce_init' ) ) {
			/* translators: 1: frontend_pricing_init 2: woocommerce_init */
			wc_doing_it_wrong( __FUNCTION__, sprintf( __( '%1$s should not be called before the %2$s action.', 'woocommerce-product-price-based-on-countries' ), 'WC_Product_Price_Based_Country::frontend_pricing_init', 'woocommerce_init' ), '3.0' );
			return;
		}

		do_action( 'wc_price_based_country_before_frontend_init' );

		// Set the current zone.
		$this->current_zone = wcpbc_get_zone_by_country();

		// Init frontend pricing.
		WCPBC_Frontend_Pricing::init();

		do_action( 'wc_price_based_country_frontend_init' );
	}

	/**
	 * Init frontend on AJAX requests. Improve compatibility with plugins which adds the "wp_ajax_nopriv_..." action on the 'init' hook.
	 */
	public function ajax_frontend_init() {
		if ( defined( 'DOING_AJAX' ) && DOING_AJAX && ! did_action( 'wc_price_based_country_frontend_init' ) ) {
			$this->frontend_init();
		}
	}

	/**
	 * Is a frontend request?.
	 *
	 * @return bool
	 */
	private function is_frontend() {
		return wcpbc_is_woocommerce_frontend() && ! defined( 'DOING_CRON' ) && ! $this->is_rest_api_for_jetpack() && ( ! is_admin() || $this->is_ajax_frontend() );
	}

	/**
	 * Check the referer type.
	 *
	 * @since 2.0.10
	 * @param string $type admin or frontend.
	 */
	private function is_referer_type( $type ) {
		$type         = is_array( $type ) ? $type : array( $type );
		$referer      = false;
		$http_referer = wp_get_referer();
		if ( $http_referer ) {
			$http_referer = str_replace( array( 'https://', 'http://', 'www.' ), '', trailingslashit( $http_referer ) );
			$admin_url    = str_replace( array( 'https://', 'http://', 'www.' ), '', trailingslashit( admin_url( '/' ) ) );
			$referer      = ( false !== strpos( $http_referer, $admin_url ) ) ? 'admin' : 'frontend';
		}

		return in_array( $referer, $type, true );
	}

	/**
	 * Is AJAX frontend request?.
	 */
	private function is_ajax_frontend() {
		$action = defined( 'DOING_AJAX' ) && DOING_AJAX && isset( $_REQUEST['action'] ) ? wc_clean( wp_unslash( $_REQUEST['action'] ) ) : false; // phpcs:ignore WordPress.Security.NonceVerification
		if ( ! $action ) {
			return false;
		}

		if ( in_array( $action, array( 'as_async_request_queue_runner' ), true ) ) {
			return false;
		}

		$is_ajax_frontend = false;

		if ( $this->is_referer_type( false ) ) {
			// Check the ajax no priv action.
			$is_ajax_frontend = has_action( 'wp_ajax_nopriv_' . $action );
		} else {
			$is_ajax_frontend = $this->is_referer_type( 'frontend' );
		}

		return apply_filters( 'wc_price_based_country_is_ajax_frontend', $is_ajax_frontend, $action );
	}

	/**
	 * Is Rest API request?.
	 *
	 * @since 2.0.0
	 * @param string $path Path to check. Default ''.
	 * @return bool
	 */
	private function is_rest_api( $path = '' ) {
		if ( empty( $_SERVER['REQUEST_URI'] ) ) {
			return false;
		}
		$path                = empty( $path ) ? $path : trailingslashit( $path );
		$rest_prefix         = trailingslashit( rest_get_url_prefix() ) . $path;
		$is_rest_api_request = ( false !== strpos( $_SERVER['REQUEST_URI'], $rest_prefix ) ); // phpcs:ignore WordPress.Security.ValidatedSanitizedInput.MissingUnslash, WordPress.Security.ValidatedSanitizedInput.InputNotSanitized

		return $is_rest_api_request;
	}

	/**
	 * Is Rest API frontend request?.
	 *
	 * @since 2.0.0
	 * @return bool
	 */
	private function is_rest_api_frontend() {
		$is_rest_api_frontend = false;

		if ( $this->is_rest_api() && $this->is_referer_type( array( 'frontend', false ) ) ) {

			$frontend_rest_api_routes = array_merge(
				array( 'wc/store/' ),
				apply_filters( 'wc_price_based_country_frontend_rest_routes', array() )
			);

			foreach ( $frontend_rest_api_routes as $route ) {
				$is_rest_api_frontend = $this->is_rest_api( $route );

				if ( $is_rest_api_frontend ) {
					break;
				}
			}
		}

		return $is_rest_api_frontend;
	}

	/**
	 * Is Rest API for Jetpack request?.
	 * Prevent issues with the WooCommerce app (android ios)
	 *
	 * @since 2.0.22
	 * @return bool
	 */
	private function is_rest_api_for_jetpack() {
		return isset( $_GET['rest_route'], $_GET['_for'] ) && 'jetpack' === $_GET['_for'] && '/wc/' === substr( $_GET['rest_route'], 0, 4 ); // phpcs:ignore WordPress.Security.NonceVerification
	}
}
